File: binary-search/search-a-2d-matrix.js
// https://leetcode.com/problems/search-a-2d-matrix/description/

class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {
        if (!matrix.length || !matrix[0].length) return false;
  
        const m = matrix.length;
        const n = matrix[0].length;
        let left = 0;
        let right = m * n - 1;
        
        while (left <= right) {
            const mid = Math.floor((left + right) / 2);
            const row = Math.floor(mid / n);
            const col = mid % n;
            const midVal = matrix[row][col];
            
            if (midVal === target) {
            return true;
            } else if (midVal < target) {
            left = mid + 1;
            } else {
            right = mid - 1;
            }
        }
        
        return false;
    }
}

File: python/binary-search/search_a_2d_matrix.py
# start off with a 2d int array called matrix
# we have another num called target that we want to find
# matrix = [[1,2,4,8],[10,11,12,13],[14,20,30,40]], target = 15 => false
# matrix = [[1,2,4,8],[10,11,12,13],[14,20,30,40]], target = 13 => true
# use the first number of index as a guide
# if matrix[i][0] > target => left mid + 1
# if matrix[i][0] < target => right mid - 1
# if left and right index are same, do binary/linear search
# if num in the list return true else return false
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        for row in matrix:
            for element in row:
                if element == target:
                    return True
        return False 